actors/evm/src/interpreter/precompiles/evm.rs

```diff
@@ -1,83 +1,64 @@
+use std::ops::RangeInclusive;
+
 use fil_actors_runtime::runtime::Runtime;
-use fvm_shared::{
-    bigint::BigUint,
-    crypto::{
-        hash::SupportedHashes,
-        signature::{SECP_SIG_LEN, SECP_SIG_MESSAGE_HASH_SIZE},
-    },
+use fvm_shared::crypto::{
+    hash::SupportedHashes,
+    signature::{SECP_SIG_LEN, SECP_SIG_MESSAGE_HASH_SIZE},
 };
 use num_traits::{One, Zero};
 use substrate_bn::{pairing_batch, AffineG1, AffineG2, Fq, Fq2, Fr, Group, Gt, G1, G2};
 use uint::byteorder::{ByteOrder, LE};
 
-use crate::interpreter::{
-    precompiles::{parameter::right_pad, PrecompileError},
-    System, U256,
-};
+use crate::interpreter::{precompiles::PrecompileError, System, U256};
 
-use super::{
-    parameter::{PaddedChunks, U256Reader},
-    PrecompileContext, PrecompileResult,
-};
+use super::{parameter::ParameterReader, PrecompileContext, PrecompileResult};
 
-lazy_static::lazy_static! {
-    pub(crate) static ref SECP256K1: BigUint = BigUint::from_bytes_be(&hex_literal::hex!("fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141"));
-}
+const SECP256K1_RANGE: RangeInclusive<U256> = U256::from_u64(2)
+    ..=U256::from_u128_words(
+        0xfffffffffffffffffffffffffffffffe,
+        0xbaaedce6af48a03bbfd25e8cd0364141,
+    );
 
-/// recover a secp256k1 pubkey from a hash, recovery byte, and a signature
-pub(super) fn ec_recover<RT: Runtime>(
-    system: &mut System<RT>,
-    input: &[u8],
-    _: PrecompileContext,
-) -> PrecompileResult {
-    let mut input_params = U256Reader::new(input);
-
-    let hash: [u8; SECP_SIG_MESSAGE_HASH_SIZE] = input_params.next_padded();
-    let recovery_byte = input_params.next_param_padded::<u8>();
-    let r = input_params.next_padded();
-    let s = input_params.next_padded();
-    let big_r = BigUint::from_bytes_be(&r);
-    let big_s = BigUint::from_bytes_be(&s);
-
-    // recovery byte is a single byte value but is represented with 32 bytes, sad
-    let v = match recovery_byte {
-        Ok(re) => {
-            if matches!(re, 27 | 28) {
-                re - 27
-            } else {
-                return Ok(Vec::new());
-            }
-        }
-        _ => return Ok(Vec::new()),
-    };
+fn ec_recover_internal<RT: Runtime>(system: &mut System<RT>, input: &[u8]) -> PrecompileResult {
+    let mut input_params = ParameterReader::new(input);
+    let hash: [u8; SECP_SIG_MESSAGE_HASH_SIZE] = input_params.read_fixed();
+    let recovery_byte: u8 = input_params.read_param()?;
+    let r: U256 = input_params.read_param()?;
+    let s: U256 = input_params.read_param()?;
 
-    let valid = if big_r <= BigUint::one() || big_s <= BigUint::one() {
-        false
-    } else {
-        big_r <= *SECP256K1 && big_s <= *SECP256K1 && (v == 0 || v == 1)
-    };
+    // Must be either 27 or 28
```

Comment:
can we `debug_assert` then?

Reply:
We check explicitly in the line below. Is that not enough?